File: q0139/q0139_recursive.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = {}
        return self.dpHelper(s, wordDict, 0, memo)
    

    # Let dpHelper(i) = True if we can build towards the end starting from index i
    def dpHelper(self, s, wordDict, i, memo):

        # Base case: if i == len(s) there is nothing left to take care, so yes we can 
        if i == len(s): return True

        # Memoized case: if we have previously checked dpHelper(i)
        if i in memo: return memo[i]

        # By default, we start with cannot build until the last index
        memo[i] = False
        # Loop through every word, if the word can be used
        # Calculate the next index j to recursively call dpHelper(j)
        for word in wordDict: 
            j = i + len(word)
            if (s[i:j] == word and self.dpHelper(s, wordDict, j, memo)): 
                memo[i] = True
                # If any word works, break immediately as there is no point exploring more
                break
                
        return memo[i]

File: q0139/q0139_recursive.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int[] memo = new int[s.length()];
        Arrays.fill(memo, -1);
        return dpHelper(s, 0, wordDict, memo);
    }

    // Let dpHelper represent if you can successfully segment the string s starting from currIdx
    private boolean dpHelper(String s, int currIdx, List<String> wordDict, int[] memo) {
        // Base case there is no need to match if the currIdx is already at the end of the string
        if (currIdx == s.length()) return true; 

        // Memoized case
        if (memo[currIdx] != -1) return memo[currIdx] == 1;

        // Recursive case
        
        for (final String word : wordDict) {
            // Check if any of the word in the dictionary works
            int nextIdx = currIdx + word.length();
            if (nextIdx <= s.length() && 
                word.equals(s.substring(currIdx, nextIdx)) &&
                dpHelper(s, nextIdx, wordDict, memo)) {
                    memo[currIdx] = 1;
                    return true;
            }
        }
        memo[currIdx] = 0;

        return false;
    }
}